src/aiida_quantumespresso/workflows/ph/base.py

```diff
@@ -103,7 +102,23 @@ def get_builder_from_protocol(cls, code, parent_folder=None, protocol=None, over
         if 'settings' in inputs['ph']:
             builder.ph['settings'] = orm.Dict(inputs['ph']['settings'])
         builder.clean_workdir = orm.Bool(inputs['clean_workdir'])
-        builder.ph['qpoints'] = qpoints
+
+        if 'qpoints' in inputs['ph']:
+            qpoints_mesh = inputs['ph'].pop('qpoints')
+            qpoints = orm.KpointsData()
+            qpoints.set_kpoints_mesh(qpoints_mesh)
+            builder.ph['qpoints'] = qpoints
+        else:
+            inputs = {
+                'structure': structure,
```

Comment:
Just wondering if we should get the `structure` from the `parent_folder`? It is a bit surprising that the `parent_folder` is optional in this method, because it is required for the `PhCalculation`, which makes sense. You cannot run a `ph.x` calculation from scratch but always need a completed `pw.x` calculation.

I am wondering then if we should not make `parent_folder` a required input and remove the `structure` input. We can then obtain the `structure` from the `parent_folder` by doing:
```python
try:
    # Use the output structure by default
    structure = parent_folder.creator.outputs.output_structure
except AttributeError:
    # Fall back on the input structure which should always exist
    structure = parent_folder.creator.inputs.structure
```

Reply:
@sphuber  Thank you! I am trying to think in the best way to organize this PR (design-wise) , because it should be linked or beneficial for `aiida-quantumespresso-ph`. `PhCalculation` has the input `parent_folder` and these are expose in `PhBaseWorkChain` so we could include this logic

Reply:
In this case, parent_folder is a expose_input from `PhCalculation`. In the `get_builder_from_protocol` there is the parent_folder so I should instead call the structure from it.